server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -267,6 +275,16 @@ public void rollGeneration() throws IOException {
     }
 
     private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOException {
+        // During primary relocation, both the old and new primary have engine created with RemoteFsTranslog and having
+        // ReplicationTracker.primaryMode() as true. However, before we perform the `internal:index/shard/replication/segments_sync`
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using startedPrimarySupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (startedPrimarySupplier.getAsBoolean() == false || syncPermit.tryAcquire(SYNC_PERMIT) == false) {
+            logger.debug("skipped uploading translog for {} {} syncPermits={}", primaryTerm, generation, syncPermit.availablePermits());
+            // NO-OP
+            return false;
+        }
```

Comment:
Why can't we simply use `ReleasableLock#tryAcquire` to make it consistent with read/write locks instead of introducing new permits?

Reply:
Will be evaluating other sempahores as well and replace with ReleasableLock. Created issue - https://github.com/opensearch-project/OpenSearch/issues/11360